tests/README.md

```diff
@@ -0,0 +1,1510 @@
+
+		$this->assertInstanceOf("ILIAS\\UI\\Component\\Button\\Factory", $f);
+		$this->assertInstanceOf
+			( "ILIAS\\UI\\Component\\Button\\Standard"
+			, $f->standard("label", "http://www.ilias.de")
+			);
+		$this->assertInstanceOf
+			( "ILIAS\\UI\\Component\\Button\\Primary"
+			, $f->primary("label", "http://www.ilias.de")
+			);
+		$this->assertInstanceOf
+			( "ILIAS\\UI\\Component\\Button\\Close"
+			, $f->close()
+			);
+		$this->assertInstanceOf
+			( "ILIAS\\UI\\Component\\Button\\Shy"
+			, $f->shy("label", "http://www.ilias.de")
+			);
+	}
+}
+```
+
+<a name="example-solution"></a>
+##### Example solution
+The proposed solution would be to remove the test entirely because they test if the new keyword works. However, if the author wishes to keep the
+tests. They could be moved into the ButtonFactoryTest class and split up in smaller more precise tests.
+As a result the developer which runs the test is now able to see which part of the factory failed.
+
+```php
+<?php
+use \ILIAS\UI\Implementation\Component\Button\Factory;
+use \ILIAS\UI\Component\Button\Standard;
+use \ILIAS\UI\Component\Button\Primary;
+use \ILIAS\UI\Component\Button\Shy;
+use \ILIAS\UI\Component\Button\Close;
+
+class ButtonFactoryTest extends AbstractFactoryTest {
+
+	private $subject;
+
+	public function setUp() {
+		parent::setUp();
+
+		$this->subject = new Factory();
+	}
+
+	// ... other parts of the test class ...
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function test_creation_of_standard_button() {
+		$this->assertInstanceOf
+				( Standard::class
+				, $this->subject->standard("label", "http://www.ilias.de")
+				);
+	}
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function test_creation_of_primary_button() {
+		$this->assertInstanceOf
+					( Primary::class
+					, $this->subject->primary("label", "http://www.ilias.de")
+					);
+	}
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function test_creation_of_Shy_button() {
+		$this->assertInstanceOf
+					( Shy::class
+					, $this->subject->shy("label", "http://www.ilias.de")
+					);
+	}
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function test_creation_of_close_button() {
+		$this->assertInstanceOf
+					( Close::class
+					, $this->subject->close("label", "http://www.ilias.de")
+					);
+	}
+}
+```
+
+
+<a name="useless-test--generic-naming"></a>
+#### Useless test / Generic naming
+Unit tests should always have an assertion of the result, because of that PHPUnit 6 started to mark such
+tests as useless. Useless test are always threaded as failed. Furthermore, the test name *test_button_label_or_glyph_only*
+is not really telling whats exactly tested.
+
+Another aspect of the whole *ButtonTest* class is that a factory is used to create concrete instances of
+the Buttons. But the factory nor the concrete subclasses of the Button class is a test subject here only the
+Button class itself. Of course to test all button instances is not really effective. However, the button can
```

Comment:
This is not true. Subject is the button component. It is more or less incidental how many concrete classes are used to implement that component. The behaviour of the component should not depend on that incident.